open-hackathon-server/src/hackathon/hazure/azure_formation.py

```diff
@@ -0,0 +1,697 @@
+        ctx = sctx.job_ctxs[sctx.current_job_index]
+        adapter = VirtualMachineAdapter(ctx.subscription_id, ctx.pem_url, host=ctx.management_host)
+
+        status = adapter.get_virtual_machine_instance_status(
+            ctx.cloud_service_name, ctx.deployment_slot, ctx.virtual_machine_name)
+
+        if not status:
+            self.log.error(
+                "azure virtual environment %d error occured while waiting for virtual machine ready"
+                % sctx.current_job_index)
+            self.__on_setup_failed(sctx)
+            return
+
+        self.log.debug(
+            "waiting for virtual machine ready, vm: %s, status: %s" %
+            (ctx.virtual_machine_name, str(status)))
+        if status == AVMStatus.READY_ROLE:
+            if ctx.vm_need_config:
+                ctx.vm_need_config = False
+                self.__config_virtual_machine(sctx)
+            else:
+                self.__setup_virtual_machine_done(sctx)
+        else:
+            self.__wait_for_virtual_machine_ready(sctx)
+
+    # callbacks will be called here
+    def __on_setup_failed(self, sctx):
+        try:
+            self.log.debug("azure virtual environment %d vm setup failed" % sctx.current_job_index)
+            ctx = sctx.job_ctxs[sctx.current_job_index]
+            ve = self.db.find_first_object_by(VirtualEnvironment, id=ctx.virtual_environment_id)
+
+            if ve:
+                ve.status = VEStatus.FAILED
+                ve.experiment.status = EStatus.FAILED
+                self.db.commit()
+        finally:
+            self.log.debug(
+                "azure virtual environment %d vm fail callback done, roll back start"
+                % sctx.current_job_index)
+            try:
+                # rollback reverse
+                self.__setup_rollback(sctx.remote_created[::-1], sctx)
+                self.log.debug("azure virtual environment %d rollback done" % sctx.current_job_index)
+            except Exception as e:
+                self.log.error(
+                    "azure virtual environment %d error while rollback: %r" %
+                    (sctx.current_job_index, str(e)))
+
+    def __setup_rollback(self, record, sctx):
+        # TODO: the rollback process should use a async way same
+        #       as previous setup process, but for convinence,
+        #       we do it in a sync way
+        for rec in record:
+            if rec.type == "cloud_service":
+                adapter = CloudServiceAdapter(sctx.subscription_id, sctx.pem_url, host=sctx.management_host)
+                adapter.delete_cloud_service(rec.name, complete=True)
+            elif rec.type == "storage_account":
+                adapter = StorageAccountAdapter(sctx.subscription_id, sctx.pem_url, host=sctx.management_host)
+                adapter.delete_storage_account(rec.name)
+            elif rec.type == "add_virtual_machine" or rec.type == "create_virtual_machine_deployment":
```

Comment:
Do we need rollback in an **exact** converse order? e.g. Will the `storage` be removed if it's still in use where VM disks are stored in?